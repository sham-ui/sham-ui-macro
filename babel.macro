/* eslint-env node */
'use strict';
const { createMacro } = require( 'babel-macros' );
const inject = require( './macros/inject' );

module.exports = createMacro( macro );

function macro( { references } ) {
    references.inject.forEach( referencePath => {
        inject( referencePath.findParent( x => x.isDecorator() ) );
    } );
}
