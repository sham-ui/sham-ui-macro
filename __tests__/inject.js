/* eslint-env node */

const path = require( 'path' );
const pluginTester = require( 'babel-plugin-tester' );
const plugin = require( 'babel-plugin-macros' );

const projectRoot = path.join( __dirname, '../../' );

expect.addSnapshotSerializer( {
    print( val ) {
        return val.split( projectRoot ).join( '<PROJECT_ROOT>/' );
    },
    test( val ) {
        return typeof val === 'string';
    }
} );

pluginTester( {
    plugin,
    snapshot: true,
    babelOptions: {
        filename: __filename,
        parserOpts: {
            plugins: [
                'decorators',
                'classProperties'
            ]
        }
    },
    tests: [ {
        title: '@inject store',
        code: `
            import { inject } from '../babel.macro';

            class FooComponent extends Template {
                @inject store;
            }
        `
    }, {
        title: '@inject(\'sham-ui:store\') store',
        code: `
            import { inject } from '../babel.macro';

            class FooComponent extends Template {
                @inject('sham-ui:store') store
            }
        `
    }, {
        title: 'with imported DI',
        code: `
            import { DI } from 'sham-ui'; 
            import { inject } from '../babel.macro';
            class Service {
                @inject api;
                @inject( 'sham-ui:store' ) uiStore;
            }
        `
    }, {
        title: 'with imported DI as alias',
        code: `
            import { DI as di } from 'sham-ui'; 
            import { inject } from '../babel.macro';
            class Service {
                @inject api;
                @inject( 'sham-ui:store' ) uiStore;
            }
        `
    } ]
} );


