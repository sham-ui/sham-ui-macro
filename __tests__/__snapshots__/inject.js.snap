// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros @inject store: @inject store 1`] = `

import { inject } from '../inject.macro';

class FooComponent extends Template {
    @inject store;
}

      ↓ ↓ ↓ ↓ ↓ ↓

class FooComponent extends Template {
    get store() {
        return this.DI.resolve('store');
    }

}

`;

exports[`macros @inject('sham-ui:store') store: @inject('sham-ui:store') store 1`] = `

import { inject } from '../inject.macro';

class FooComponent extends Template {
    @inject('sham-ui:store') store
}

      ↓ ↓ ↓ ↓ ↓ ↓

class FooComponent extends Template {
    get store() {
        return this.DI.resolve('sham-ui:store');
    }

}

`;

exports[`macros with imported DI as alias: with imported DI as alias 1`] = `

import { DI as di } from 'sham-ui'; 
import { inject } from '../inject.macro';
class Service {
    @inject api;
    @inject( 'sham-ui:store' ) uiStore;
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { DI as di } from 'sham-ui';

class Service {
    get api() {
        return this.DI.resolve('api');
    }

    get uiStore() {
        return this.DI.resolve('sham-ui:store');
    }

}

`;

exports[`macros with imported DI: with imported DI 1`] = `

import { DI } from 'sham-ui'; 
import { inject } from '../inject.macro';
class Service {
    @inject api;
    @inject( 'sham-ui:store' ) uiStore;
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { DI } from 'sham-ui';

class Service {
    get api() {
        return this.DI.resolve('api');
    }

    get uiStore() {
        return this.DI.resolve('sham-ui:store');
    }

}

`;
